// graphql-compose-mongoose/lib/resolvers/findMany.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findMany = void 0;
const helpers_1 = require("./helpers");
const beforeQueryHelper_1 = require("./helpers/beforeQueryHelper");
function findMany(model, tc, opts) {
    if (!model || !model.modelName || !model.schema) {
        throw new Error('First arg for Resolver findMany() should be instance of Mongoose Model.');
    }
    if (!tc || tc.constructor.name !== 'ObjectTypeComposer') {
        throw new Error('Second arg for Resolver findMany() should be instance of ObjectTypeComposer.');
    }
    const aliases = (0, helpers_1.prepareNestedAliases)(model.schema);
    const aliasesReverse = (0, helpers_1.prepareAliasesReverse)(model.schema);
    return tc.schemaComposer.createResolver({
        type: tc.NonNull.List.NonNull,
        name: 'findMany',
        kind: 'query',
        args: Object.assign(Object.assign(Object.assign(Object.assign({}, (0, helpers_1.filterHelperArgs)(tc, model, Object.assign({ prefix: 'FilterFindMany', suffix: `${(opts === null || opts === void 0 ? void 0 : opts.suffix) || ''}Input` }, opts === null || opts === void 0 ? void 0 : opts.filter))), (0, helpers_1.skipHelperArgs)()), (0, helpers_1.limitHelperArgs)(Object.assign({}, opts === null || opts === void 0 ? void 0 : opts.limit))), (0, helpers_1.sortHelperArgs)(tc, model, Object.assign({ sortTypeName: `SortFindMany${tc.getTypeName()}${(opts === null || opts === void 0 ? void 0 : opts.suffix) || ''}Input` }, opts === null || opts === void 0 ? void 0 : opts.sort))),
        resolve: ((resolveParams) => __awaiter(this, void 0, void 0, function* () {
            resolveParams.query = model.find();
            resolveParams.model = model;
            (0, helpers_1.filterHelper)(resolveParams, aliases);
            (0, helpers_1.skipHelper)(resolveParams);
            (0, helpers_1.limitHelper)(resolveParams);
            (0, helpers_1.sortHelper)(resolveParams);
            (0, helpers_1.projectionHelper)(resolveParams, aliases);
            if (opts === null || opts === void 0 ? void 0 : opts.lean) {
                const result = (yield (0, beforeQueryHelper_1.beforeQueryHelperLean)(resolveParams)) || [];
                return Array.isArray(result) && aliasesReverse
                    ? result.map((r) => (0, helpers_1.replaceAliases)(r, aliasesReverse))
                    : result;
            }
            else {
                return (0, beforeQueryHelper_1.beforeQueryHelper)(resolveParams) || [];
            }
        })),
    });
}
exports.findMany = findMany;
