// graphql-compose-mongoose/lib/resolvers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolverFactory = void 0;
const count_1 = require("./count");
const findById_1 = require("./findById");
const findByIds_1 = require("./findByIds");
const findMany_1 = require("./findMany");
const findOne_1 = require("./findOne");
const createMany_1 = require("./createMany");
const createOne_1 = require("./createOne");
const updateById_1 = require("./updateById");
const updateMany_1 = require("./updateMany");
const updateOne_1 = require("./updateOne");
const removeById_1 = require("./removeById");
const removeMany_1 = require("./removeMany");
const removeOne_1 = require("./removeOne");
const dataLoader_1 = require("./dataLoader");
const dataLoaderMany_1 = require("./dataLoaderMany");
const pagination_1 = require("./pagination");
const connection_1 = require("./connection");
exports.resolverFactory = {
    count: count_1.count,
    findById: findById_1.findById,
    findByIds: findByIds_1.findByIds,
    findOne: findOne_1.findOne,
    findMany: findMany_1.findMany,
    dataLoader: dataLoader_1.dataLoader,
    dataLoaderMany: dataLoaderMany_1.dataLoaderMany,
    createOne: createOne_1.createOne,
    createMany: createMany_1.createMany,
    updateById: updateById_1.updateById,
    updateOne: updateOne_1.updateOne,
    updateMany: updateMany_1.updateMany,
    removeById: removeById_1.removeById,
    removeOne: removeOne_1.removeOne,
    removeMany: removeMany_1.removeMany,
    pagination: pagination_1.pagination,
    connection: connection_1.connection,
};
