// graphql-compose-mongoose/lib/resolvers/removeById.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeById = void 0;
const graphql_compose_1 = require("graphql-compose");
const findById_1 = require("./findById");
const errorCatcher_1 = require("./helpers/errorCatcher");
const payloadRecordId_1 = require("./helpers/payloadRecordId");
function removeById(model, tc, opts) {
    if (!model || !model.modelName || !model.schema) {
        throw new Error('First arg for Resolver removeById() should be instance of Mongoose Model.');
    }
    if (!tc || tc.constructor.name !== 'ObjectTypeComposer') {
        throw new Error('Second arg for Resolver removeById() should be instance of ObjectTypeComposer.');
    }
    const findByIdResolver = (0, findById_1.findById)(model, tc);
    const outputTypeName = `RemoveById${tc.getTypeName()}${(opts === null || opts === void 0 ? void 0 : opts.suffix) || ''}Payload`;
    const outputType = tc.schemaComposer.getOrCreateOTC(outputTypeName, (t) => {
        t.setFields(Object.assign(Object.assign({}, (0, payloadRecordId_1.payloadRecordId)(tc, opts === null || opts === void 0 ? void 0 : opts.recordId)), { record: {
                type: tc,
                description: 'Removed document',
            } }));
    });
    const resolver = tc.schemaComposer.createResolver({
        name: 'removeById',
        kind: 'mutation',
        description: 'Remove one document: ' +
            '1) Retrieve one document and remove with hooks via findByIdAndRemove. ' +
            '2) Return removed document.',
        type: outputType,
        args: {
            _id: tc.hasField('_id') ? (0, graphql_compose_1.toInputType)(tc.getFieldTC('_id')).NonNull : 'MongoID!',
        },
        resolve: ((resolveParams) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            const _id = (_a = resolveParams === null || resolveParams === void 0 ? void 0 : resolveParams.args) === null || _a === void 0 ? void 0 : _a._id;
            if (!_id) {
                throw new Error(`${tc.getTypeName()}.removeById resolver requires args._id value`);
            }
            let doc = yield findByIdResolver.resolve(Object.assign(Object.assign({}, resolveParams), { projection: {} }));
            if (resolveParams.beforeRecordMutate) {
                doc = yield resolveParams.beforeRecordMutate(doc, resolveParams);
            }
            if (doc) {
                yield doc.remove();
                return {
                    record: doc,
                };
            }
            return null;
        })),
    });
    if (!(opts === null || opts === void 0 ? void 0 : opts.disableErrorField)) {
        (0, errorCatcher_1.addErrorCatcherField)(resolver);
    }
    return resolver;
}
exports.removeById = removeById;
