// graphql-compose-mongoose/lib/resolvers/removeMany.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeMany = void 0;
const helpers_1 = require("./helpers");
const beforeQueryHelper_1 = require("./helpers/beforeQueryHelper");
const errorCatcher_1 = require("./helpers/errorCatcher");
function removeMany(model, tc, opts) {
    if (!model || !model.modelName || !model.schema) {
        throw new Error('First arg for Resolver removeMany() should be instance of Mongoose Model.');
    }
    if (!tc || tc.constructor.name !== 'ObjectTypeComposer') {
        throw new Error('Second arg for Resolver removeMany() should be instance of ObjectTypeComposer.');
    }
    const outputTypeName = `RemoveMany${tc.getTypeName()}${(opts === null || opts === void 0 ? void 0 : opts.suffix) || ''}Payload`;
    const outputType = tc.schemaComposer.getOrCreateOTC(outputTypeName, (t) => {
        t.addFields({
            numAffected: {
                type: 'Int',
                description: 'Affected documents number',
            },
        });
    });
    const aliases = (0, helpers_1.prepareNestedAliases)(model.schema);
    const resolver = tc.schemaComposer.createResolver({
        name: 'removeMany',
        kind: 'mutation',
        description: 'Remove many documents without returning them: ' +
            'Use Query.remove mongoose method. ' +
            'Do not apply mongoose defaults, setters, hooks and validation. ',
        type: outputType,
        args: Object.assign(Object.assign({}, (0, helpers_1.filterHelperArgs)(tc, model, Object.assign({ prefix: 'FilterRemoveMany', suffix: `${(opts === null || opts === void 0 ? void 0 : opts.suffix) || ''}Input`, isRequired: true }, opts === null || opts === void 0 ? void 0 : opts.filter))), (0, helpers_1.limitHelperArgs)(Object.assign({}, opts === null || opts === void 0 ? void 0 : opts.limit))),
        resolve: ((resolveParams) => __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const filterData = (_a = resolveParams === null || resolveParams === void 0 ? void 0 : resolveParams.args) === null || _a === void 0 ? void 0 : _a.filter;
            if (!(typeof filterData === 'object') || Object.keys(filterData).length === 0) {
                throw new Error(`${tc.getTypeName()}.removeMany resolver requires at least one value in args.filter`);
            }
            resolveParams.query = model.find();
            resolveParams.model = model;
            (0, helpers_1.filterHelper)(resolveParams, aliases);
            if (resolveParams.query.deleteMany) {
                resolveParams.query = resolveParams.query.deleteMany();
            }
            else {
                resolveParams.query = resolveParams.query.remove();
            }
            const res = yield (0, beforeQueryHelper_1.beforeQueryHelper)(resolveParams);
            if (typeof res.deletedCount === 'number') {
                return {
                    numAffected: res.deletedCount,
                };
            }
            else if (res.ok) {
                return {
                    numAffected: res.n,
                };
            }
            else if ((_b = res === null || res === void 0 ? void 0 : res.result) === null || _b === void 0 ? void 0 : _b.ok) {
                return {
                    numAffected: res.result.n,
                };
            }
            throw new Error(JSON.stringify(res));
        })),
    });
    if (!(opts === null || opts === void 0 ? void 0 : opts.disableErrorField)) {
        (0, errorCatcher_1.addErrorCatcherField)(resolver);
    }
    return resolver;
}
exports.removeMany = removeMany;
