// graphql-compose-mongoose/lib/resolvers/updateOne.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateOne = void 0;
const helpers_1 = require("./helpers");
const findOne_1 = require("./findOne");
const errorCatcher_1 = require("./helpers/errorCatcher");
const validate_1 = require("./helpers/validate");
const payloadRecordId_1 = require("./helpers/payloadRecordId");
function updateOne(model, tc, opts) {
    if (!model || !model.modelName || !model.schema) {
        throw new Error('First arg for Resolver updateOne() should be instance of Mongoose Model.');
    }
    if (!tc || tc.constructor.name !== 'ObjectTypeComposer') {
        throw new Error('Second arg for Resolver updateOne() should be instance of ObjectTypeComposer.');
    }
    const findOneResolver = (0, findOne_1.findOne)(model, tc, opts);
    const outputTypeName = `UpdateOne${tc.getTypeName()}${(opts === null || opts === void 0 ? void 0 : opts.suffix) || ''}Payload`;
    const outputType = tc.schemaComposer.getOrCreateOTC(outputTypeName, (t) => {
        t.setFields(Object.assign(Object.assign({}, (0, payloadRecordId_1.payloadRecordId)(tc, opts === null || opts === void 0 ? void 0 : opts.recordId)), { record: {
                type: tc,
                description: 'Updated document',
            } }));
    });
    const resolver = tc.schemaComposer.createResolver({
        name: 'updateOne',
        kind: 'mutation',
        description: 'Update one document: ' +
            '1) Retrieve one document via findOne. ' +
            '2) Apply updates to mongoose document. ' +
            '3) Mongoose applies defaults, setters, hooks and validation. ' +
            '4) And save it.',
        type: outputType,
        args: Object.assign(Object.assign(Object.assign(Object.assign({}, (0, helpers_1.recordHelperArgs)(tc, Object.assign({ prefix: 'UpdateOne', suffix: `${(opts === null || opts === void 0 ? void 0 : opts.suffix) || ''}Input`, removeFields: ['id', '_id'], isRequired: true, allFieldsNullable: true }, opts === null || opts === void 0 ? void 0 : opts.record))), (0, helpers_1.filterHelperArgs)(tc, model, Object.assign({ prefix: 'FilterUpdateOne', suffix: `${(opts === null || opts === void 0 ? void 0 : opts.suffix) || ''}Input` }, opts === null || opts === void 0 ? void 0 : opts.filter))), (0, helpers_1.sortHelperArgs)(tc, model, Object.assign({ sortTypeName: `SortUpdateOne${tc.getTypeName()}${(opts === null || opts === void 0 ? void 0 : opts.suffix) || ''}Input` }, opts === null || opts === void 0 ? void 0 : opts.sort))), (0, helpers_1.skipHelperArgs)()),
        resolve: ((resolveParams) => __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const recordData = (_a = resolveParams === null || resolveParams === void 0 ? void 0 : resolveParams.args) === null || _a === void 0 ? void 0 : _a.record;
            const filterData = (_b = resolveParams === null || resolveParams === void 0 ? void 0 : resolveParams.args) === null || _b === void 0 ? void 0 : _b.filter;
            if (!(typeof filterData === 'object') || Object.keys(filterData).length === 0) {
                return Promise.reject(new Error(`${tc.getTypeName()}.updateOne resolver requires at least one value in args.filter`));
            }
            let doc = yield findOneResolver.resolve(Object.assign(Object.assign({}, resolveParams), { projection: {} }));
            if (resolveParams.beforeRecordMutate) {
                doc = yield resolveParams.beforeRecordMutate(doc, resolveParams);
            }
            if (!doc)
                return null;
            doc.set(recordData);
            yield (0, validate_1.validateAndThrow)(doc);
            yield doc.save({ validateBeforeSave: false });
            return {
                record: doc,
            };
        })),
    });
    if (!(opts === null || opts === void 0 ? void 0 : opts.disableErrorField)) {
        (0, errorCatcher_1.addErrorCatcherField)(resolver);
    }
    return resolver;
}
exports.updateOne = updateOne;
