// graphql-compose-mongoose/lib/resolvers/helpers/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MergeAbleHelperArgsOpts = void 0;
const filter_1 = require("./filter");
const limit_1 = require("./limit");
const record_1 = require("./record");
__exportStar(require("./aliases"), exports);
__exportStar(require("./filter"), exports);
__exportStar(require("./limit"), exports);
__exportStar(require("./projection"), exports);
__exportStar(require("./record"), exports);
__exportStar(require("./skip"), exports);
__exportStar(require("./sort"), exports);
exports.MergeAbleHelperArgsOpts = {
    sort: 'boolean',
    skip: 'boolean',
    limit: (0, limit_1.getLimitHelperArgsOptsMap)(),
    filter: (0, filter_1.getFilterHelperArgOptsMap)(),
    record: (0, record_1.getRecordHelperArgsOptsMap)(),
    records: (0, record_1.getRecordHelperArgsOptsMap)(),
};
